reference/reference.go

```diff
@@ -190,45 +211,153 @@ func ParseNamed(s string) (Named, error) {
 	}
 	named, isNamed := ref.(Named)
 	if !isNamed {
-		return nil, fmt.Errorf("reference %s has no name", ref.String())
+		return nil, ErrNameRequired
+	}
+	return named, nil
+}
+
+// ParseNameOnly is a strict version of ParseNamed that only accepts
+// references that are valid names.
+func ParseNameOnly(name string) (Named, error) {
+	named, err := ParseNamed(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !NamedOnly(named) {
+		return nil, ErrNameOnlyAllowed
 	}
+
 	return named, nil
 }
 
-// WithName returns a named object representing the given string. If the input
-// is invalid ErrReferenceInvalidFormat will be returned.
-func WithName(name string) (Named, error) {
+// NamedOnly returns true if reference only contains a repo name and not
+// other modifiers.
+func NamedOnly(ref Named) bool {
+	return ref.Name() == ref.String()
+}
+
+// NameOnly drops other reference information and only retains the name.
+func NameOnly(ref Named) Named {
+	named, ok := ref.(repository)
+	if ok {
+		return named
+	}
+
+	return repository(ref.Name())
+}
+
+// WithName replaces or sets Name() on the reference with the provided value
+// of name. This is useful for specifying a name for a digest only reference.
+//
+// If the target of named is from another package, any data on the backing
+// type will be lost and replaced with an instance from this package. Simply
+// put, the resulting type will only include information available on the
+// interface.
+func WithName(ref Reference, name string) (Named, error) {
+	if name == "" {
+		return nil, ErrNameEmpty
+	}
+
 	if len(name) > NameTotalLengthMax {
 		return nil, ErrNameTooLong
 	}
+
 	if !anchoredNameRegexp.MatchString(name) {
 		return nil, ErrReferenceInvalidFormat
 	}
-	return repository(name), nil
+
+	if ref == nil {
+		return repository(name), nil
+	}
+
+	switch v := ref.(type) {
+	case repository:
+		return v, nil
+	case reference:
+		v.name = name
+		return v, nil
+	case digestReference:
+		return canonicalReference{
+			name:   name,
+			digest: v.Digest(),
+		}, nil
+	case taggedReference:
+		v.name = name
+		return v, nil
+	case canonicalReference:
+		v.name = name
+		return v, nil
+	}
```

Comment:
It looks like this won't play nicely without backing types defined outside this package. This seems like an oversight, because the godoc comment says implies it works with other types.